File: app/src/main/java/com/theone/music/ui/fragment/signer/SingerFragment.java
package com.theone.music.ui.fragment.signer;//  ┏┓　　　┏┓
//┏┛┻━━━┛┻┓
//┃　　　　　　　┃
//┃　　　━　　　┃
//┃　┳┛　┗┳　┃
//┃　　　　　　　┃
//┃　　　┻　　　┃
//┃　　　　　　　┃
//┗━┓　　　┏━┛
//    ┃　　　┃                  神兽保佑
//    ┃　　　┃                  永无BUG！
//    ┃　　　┗━━━┓
//    ┃　　　　　　　┣┓
//    ┃　　　　　　　┏┛
//    ┗┓┓┏━┳┓┏┛
//      ┃┫┫　┃┫┫
//      ┗┻┛　┗┻┛

import android.util.SparseArray;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;

import com.theone.music.BR;
import com.theone.music.R;
import com.theone.music.data.model.Singer;
import com.theone.music.databinding.FragmentSingerBinding;
import com.theone.music.ui.fragment.BaseFragment;
import com.theone.music.viewmodel.SingerViewModel;
import com.theone.mvvm.core.app.ext.LoaderExtKt;
import com.theone.mvvm.ext.qmui.QMUITopBarLayoutExtKt;

import java.util.List;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;

/**
 * @author The one
 * @date 2022-04-11 14:12
 * @describe TODO
 * @email 625805189@qq.com
 * @remark
 */
public class SingerFragment extends BaseFragment<SingerViewModel, FragmentSingerBinding> {

    @Nullable
    @Override
    public View loaderRegisterView() {
        return getContentView();
    }

    @Override
    public void initView(@NonNull View view) {
        QMUITopBarLayoutExtKt.setTitleWitchBackBtn(this, "歌手");
        onPageReLoad();
    }

    @Override
    public void onPageReLoad() {
        LoaderExtKt.showLoadingPage(this, "加载中");
        getViewModel().requestServer();
    }

    @Override
    public void createObserver() {
        getViewModel().getResponseLiveData().observe(this, new Observer<List<? extends Singer>>() {
            @Override
            public void onChanged(List<? extends Singer> singers) {

            }
        });
        getViewModel().getErrorLiveData().observe(this, new Observer<String>() {
            @Override
            public void onChanged(String s) {
                LoaderExtKt.showErrorPage(SingerFragment.this,
                        s,
                        R.drawable.status_loading_view_loading_fail,
                        new Function1<View, Unit>() {
                            @Override
                            public Unit invoke(View view) {
                                onPageReLoad();
                                return null;
                            }
                        });
            }
        });
    }

    @Override
    public void applyBindingParams(@NonNull SparseArray<Object> sparseArray) {
        sparseArray.append(BR.host,this);
    }
}

File: app/src/main/java/com/theone/music/ui/fragment/signer/SingerFragment.kt
package com.theone.music.ui.fragment.signer

import android.util.TypedValue
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.qmuiteam.qmui.kotlin.wrapContent
import com.qmuiteam.qmui.layout.QMUIFrameLayout
import com.theone.common.ext.dp2px
import com.theone.common.ext.getDrawable
import com.theone.music.R
import com.theone.music.app.util.ColorUtil
import com.theone.music.data.model.Singer
import com.theone.music.databinding.FragmentSingerBinding
import com.theone.music.viewmodel.SingerViewModel
import com.theone.mvvm.core.app.ext.showErrorPage
import com.theone.mvvm.core.app.ext.showLoadingPage
import com.theone.mvvm.core.app.ext.showSuccessPage
import com.theone.mvvm.core.base.fragment.BaseCoreFragment
import com.theone.mvvm.ext.qmui.setTitleWitchBackBtn

//  ┏┓　　　┏┓
//┏┛┻━━━┛┻┓
//┃　　　　　　　┃
//┃　　　━　　　┃
//┃　┳┛　┗┳　┃
//┃　　　　　　　┃
//┃　　　┻　　　┃
//┃　　　　　　　┃
//┗━┓　　　┏━┛
//    ┃　　　┃                  神兽保佑
//    ┃　　　┃                  永无BUG！
//    ┃　　　┗━━━┓
//    ┃　　　　　　　┣┓
//    ┃　　　　　　　┏┛
//    ┗┓┓┏━┳┓┏┛
//      ┃┫┫　┃┫┫
//      ┗┻┛　┗┻┛
/**
 * @author The one
 * @date 2022-04-07 09:09
 * @describe TODO
 * @email 625805189@qq.com
 * @remark
 */
class SingerFragment : BaseCoreFragment<SingerViewModel, FragmentSingerBinding>() {

    override fun loaderRegisterView(): View = getContentView()

    override fun initView(root: View) {
        setTitleWitchBackBtn("歌手")
        onPageReLoad()
    }

    override fun onPageReLoad() {
        showLoadingPage()
        getViewModel().requestServer()
    }

    override fun createObserver() {
        getViewModel().getResponseLiveData().observe(this) {
            layoutSinger(it)
            showSuccessPage()
        }
        getViewModel().getErrorLiveData().observe(this) {
            showErrorPage(it) {
                onPageReLoad()
            }
        }
    }

    private fun layoutSinger(singers: List<Singer>) {
        getDataBinding().floatLayout
            .run {
                removeAllViews()
                for (singer in singers) {
                    val layoutParams = ViewGroup.LayoutParams(wrapContent, wrapContent)
                    val container = QMUIFrameLayout(context)
                    val space = context.dp2px(10)
                    container.setPadding(0, 0, space, space)
                    val tag = TextView(context)
                    val padding = context.dp2px(4)
                    val padding2 = padding * 2
                    tag.run {
                        setPadding(padding2, padding, padding2, padding)
                        setTextSize(TypedValue.COMPLEX_UNIT_SP, 15f)
                        maxLines = 1
                        setTextColor(ColorUtil.randomColor())
                        text = singer.name
                        background = getDrawable(
                            context,
                            R.drawable.tree_tag_bg
                        )
                    }
                    container.addView(tag, layoutParams)
                    container.setOnClickListener {
                        startFragment(SignerSearchFragment.newInstance(singer))
                    }
                    addView(container, layoutParams)
                }
            }
    }

}